File: app/src/main/java/com/example/android/myaddressbook/ContactsAdapter.java
package com.example.android.myaddressbook;

import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public class ContactsAdapter extends RecyclerView.Adapter<ContactViewHolder> {

    private AppCompatActivity activity;

    private ArrayList<Contact> contacts;

    public ContactsAdapter(AppCompatActivity activity, ArrayList<Contact> contacts) {
        this.activity = activity;
        this.contacts = contacts;
    }

    @Override public ContactViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ContactViewHolder(activity, parent);
    }

    @Override public void onBindViewHolder(ContactViewHolder holder, final int position) {
        holder.bind(contacts.get(position));
    }

    @Override public int getItemCount() {
        return contacts.size();
    }

    public ArrayList<Contact> getContacts() {
        return contacts;
    }

    public void replaceContacts(List<Contact> newContacts) {
        contacts.clear();
        contacts.addAll(newContacts);
        notifyDataSetChanged();
    }

    public void saveContact(Contact editedContact, int contactPosition) {
        if (contactPosition == -1) {
            contacts.add(editedContact);
            notifyItemInserted(contacts.size());
        } else {
            contacts.set(contactPosition, editedContact);
            notifyItemChanged(contactPosition);
        }
    }
}

File: app/src/main/java/com/example/android/myaddressbook/ContactsAdapter.kt
package com.example.android.myaddressbook

import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import java.util.*

class ContactsAdapter(
        private val activity: AppCompatActivity,
        val contacts: ArrayList<Contact>
) : RecyclerView.Adapter<ContactViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ContactViewHolder(activity, parent)

    override fun onBindViewHolder(holder: ContactViewHolder, position: Int) {
        holder.bind(contacts[position])
    }

    override fun getItemCount() = contacts.size

    fun replaceContacts(newContacts: List<Contact>) {
        contacts.clear()
        contacts += newContacts
        notifyDataSetChanged()
    }

    fun saveContact(editedContact: Contact, contactPosition: Int) {
        if (contactPosition == -1) {
            contacts += editedContact
            notifyItemInserted(contacts.size)
        } else {
            contacts[contactPosition] = editedContact
            notifyItemChanged(contactPosition)
        }
    }
}
